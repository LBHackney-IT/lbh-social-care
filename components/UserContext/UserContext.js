import { createContext, useContext, useEffect, useState } from 'react';
import { useRouter } from 'next/router';

import { AUTH_WHITELIST } from 'utils/auth';
import { getUser } from 'utils/api/checkAuth';
import Layout from 'components/Layout';
import Spinner from 'components/Spinner/Spinner';

export const UserContext = createContext({
  user: undefined,
  setUser: () => {},
});

export const AuthProvider = ({ children }) => {
  const [user, setUser] = useState();
  const [redirect, setRedirect] = useState();
  const [isLoading, setLoading] = useState(true);
  const router = useRouter();
  useEffect(() => {
    const initializeAuth = async () => {
      try {
        const data = getUser();
        setUser(data);
      } catch (e) {
        setUser();
        setRedirect(e.response.status === 403 ? '/access-denied' : '/login');
      }
      setLoading(false);
    };
    initializeAuth();
  }, []);
  if (isLoading || !router.isReady) {
    return (
      <Layout>
        <Spinner />
      </Layout>
    );
  }
  if (
    typeof window !== 'undefined' &&
    AUTH_WHITELIST.includes(router.pathname) === Boolean(user)
  ) {
    router.push(redirect || location);
    return <></>;
  }
  return (
    <UserContext.Provider
      value={{
        user,
        isLoading,
        setUser,
        redirect,
      }}
    >
      {children}
    </UserContext.Provider>
  );
};

export function useAuth() {
  const context = useContext(UserContext);
  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  return context;
}
