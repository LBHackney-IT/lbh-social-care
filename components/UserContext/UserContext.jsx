import { createContext, useContext } from 'react';
import { useRouter } from 'next/router';

import { shouldRedirect } from 'utils/auth';
import { isBrowser } from 'utils/ssr';

export const UserContext = createContext({
  user: undefined,
  setUser: () => {},
});

export const AuthProvider = ({ children, user }) => {
  const { pathname, push } = useRouter();
  const redirect = isBrowser() && shouldRedirect(pathname, user);
  if (redirect) {
    push(redirect);
    return <></>;
  }
  return (
    <UserContext.Provider
      value={{
        user,
      }}
    >
      {children}
    </UserContext.Provider>
  );
};

export function useAuth() {
  const context = useContext(UserContext);
  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  return context;
}
