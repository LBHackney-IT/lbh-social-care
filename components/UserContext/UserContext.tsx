import { createContext, useContext, ReactNode, ReactElement } from 'react';
import { useRouter } from 'next/router';

import { shouldRedirect } from 'utils/auth';
import { isBrowser } from 'utils/ssr';

import type { User } from 'types';

interface Props {
  children: ReactNode;
  user?: User;
}

export const UserContext = createContext<{ user?: User }>({
  user: undefined,
});

export const AuthProvider = ({ children, user }: Props): ReactElement => {
  const { pathname, push } = useRouter();
  const redirect = isBrowser() && shouldRedirect(pathname, user);
  if (redirect) {
    push(redirect);
    return <></>;
  }
  return (
    <UserContext.Provider
      value={{
        user,
      }}
    >
      {children}
    </UserContext.Provider>
  );
};

export const useAuth = (): { user?: User } => {
  const context = useContext(UserContext);
  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  return context;
};
