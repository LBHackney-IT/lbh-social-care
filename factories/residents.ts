import { Factory } from 'fishery';

import { Resident, ExtendedResident } from 'types';

export const legacyResidentFactory = Factory.define<Resident>(
  ({ sequence }) => ({
    mosaicId: sequence,
    dateOfBirth: '2020-11-13',
    firstName: 'Foo',
    lastName: 'Bar',
    nhsNumber: '12345',
    ageContext: 'A',
    gender: 'F',
  })
);

export const residentFactory = Factory.define<ExtendedResident>(
  ({ sequence }) => ({
    personId: sequence,
    dateOfBirth: '2020-11-13',
    firstName: 'Foo',
    lastName: 'Bar',
    nhsNumber: 12345,
    contextFlag: 'A',
    gender: 'F',
    createdBy: 'foo@bar.com',
    otherNames: [],
    phoneNumbers: [],
  })
);

export const mockedLegacyResident = legacyResidentFactory.build();
export const mockedResident = residentFactory.build();
