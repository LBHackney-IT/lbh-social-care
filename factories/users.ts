import { Factory } from 'fishery';

import { User } from 'types';

export const userFactory = Factory.define<User>(() => ({
  name: 'foo',
  hasAdminPermissions: true,
  hasChildrenPermissions: true,
  hasAdultPermissions: true,
  email: 'foo@bar.com',
  permissionFlag: 'A',
  isAuthorised: true,
}));

export const mockedUser = userFactory.build();
