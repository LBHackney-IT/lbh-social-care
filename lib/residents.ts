import axios from 'axios';

import type { LegacyResident, Resident, ResidentsAPI, AgeContext } from 'types';

const { ENDPOINT_API, AWS_KEY } = process.env;

const headers = { 'x-api-key': AWS_KEY };

interface ResidentBE extends Omit<LegacyResident, 'restricted'> {
  restricted: 'Y' | 'N';
  addressList: Array<{
    displayAddressFlag: 'Y' | 'N';
    addressLine1: string;
    postCode: string;
    uprn: string;
  }>;
}

const sanitiseResidentData = (residents: ResidentBE[]): LegacyResident[] =>
  residents?.map(
    ({ addressList, ...resident }: ResidentBE): LegacyResident => {
      const address = addressList?.find(
        ({ displayAddressFlag }) => displayAddressFlag === 'Y'
      );
      return {
        ...resident,
        restricted: resident.restricted === 'Y',
        address: address && {
          address: address.addressLine1,
          postcode: address.postCode,
          uprn: address.uprn,
        },
      };
    }
  );

export const getResidents = async (
  params: Record<string, unknown>
): Promise<ResidentsAPI> => {
  const { data } = await axios.get(`${ENDPOINT_API}/residents`, {
    headers,
    params,
  });
  return { ...data, residents: sanitiseResidentData(data.residents) };
};

export const getResident = async (
  personId: number,
  params: { context_flag: AgeContext }
): Promise<Resident> => {
  const { data } = await axios.get(`${ENDPOINT_API}/residents/${personId}`, {
    headers,
    params,
  });
  return { ...data, restricted: data.restricted === 'Y' };
};

export const normalisePhoneInput = (formData: {
  phoneNumbers?: Record<string, string>[];
  [key: string]: unknown;
}): Record<string, unknown> => ({
  ...formData,
  phoneNumbers: formData.phoneNumbers
    ? formData.phoneNumbers.map((phone) => ({
        ...phone,
        type: phone.type === '' ? 'main' : phone.type,
      }))
    : null,
});

export const addResident = async (
  formData: Record<string, unknown>
): Promise<ResidentsAPI> => {
  const { data } = await axios.post(
    `${ENDPOINT_API}/residents`,
    normalisePhoneInput(formData),
    {
      headers: { ...headers, 'Content-Type': 'application/json' },
    }
  );
  return data;
};

export const updateResident = async (
  formData: Record<string, unknown>
): Promise<ResidentsAPI> => {
  const { data } = await axios.patch(
    `${ENDPOINT_API}/residents`,
    normalisePhoneInput(formData),
    {
      headers: { ...headers, 'Content-Type': 'application/json' },
    }
  );
  return data;
};
