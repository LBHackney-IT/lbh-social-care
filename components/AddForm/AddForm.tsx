import React, { useState } from 'react';

import Button from 'components/Button/Button';
import { useAuth } from 'components/UserContext/UserContext';
import { Autocomplete } from 'components/Form/Autocomplete/Autocomplete';
import { populateChildForm } from 'utils/populate';
import ADULT_FORMS from 'data/googleForms/adultForms';
import CHILD_FORMS from 'data/googleForms/childForms';
import { Resident, User } from 'types';

const AddForm = ({ person }: { person: Resident }): React.ReactElement => {
  const { user } = useAuth() as { user: User };
  const [url, setUrl] = useState<string>();
  const ageContext = person && person.contextFlag;
  const forms = ageContext === 'C' ? CHILD_FORMS : ADULT_FORMS;
  const internalForms =
    ageContext === 'A'
      ? [
          {
            text: 'Case Note Recording',
            value: `/people/${person.id}/records/case-notes-recording`,
          },
          // {
          //   text: 'Conversation 3',
          //   value: `/people/${person.id}/records/conversation-3`,
          // },
        ]
      : [];
  const betaForms = user.hasDevPermissions
    ? [
        {
          text: 'Warning Note',
          value: `/people/${person.id}/warning-notes/add`,
        },
      ]
    : [];
  return (
    <>
      <div className="lbh-form-group govuk-form-group">
        <Autocomplete
          name="formList"
          options={[...internalForms, ...forms, ...betaForms]}
          label="Choose a form"
          placeholder="Select or type form name"
          onChange={(value) => setUrl(value as string)}
        />
      </div>
      <Button
        label="Load form"
        route={
          ageContext === 'C'
            ? `${url}${populateChildForm(
                person.firstName,
                person.lastName,
                person.id,
                user.name,
                url
              )}`
            : url
        }
        internalQuery={`?id=${person.id}`}
      />
    </>
  );
};

export default AddForm;
