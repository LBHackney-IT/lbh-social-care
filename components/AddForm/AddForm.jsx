import { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import LinkButton from 'components/LinkButton/LinkButton';
import ErrorMessage from 'components/ErrorMessage/ErrorMessage';
import Spinner from 'components/Spinner/Spinner';
import { useAuth } from 'components/UserContext/UserContext';
import { Select } from 'components/Form';
import { getResident } from 'utils/api/residents';
import { populateChildForm } from 'utils/populate';
import ADULT_CATE from 'data/googleForms/adultCategories';
import ADULT_FORMS from 'data/googleForms/adultForms';
import CHILD_CATE from 'data/googleForms/childCategories';
import CHILD_FORMS from 'data/googleForms/childForms';

const AddForm = ({ id }) => {
  const { user } = useAuth();
  const [loading, setLoading] = useState(true);
  const [person, setPerson] = useState();
  const [error, setError] = useState();
  const [url, setUrl] = useState();
  const [categoryValue, setCategoryValue] = useState();
  const ageContext = person && person.ageContext;
  const category = ageContext === 'C' ? CHILD_CATE : ADULT_CATE;
  const forms = ageContext === 'C' ? CHILD_FORMS : ADULT_FORMS;

  const getPerson = async () => {
    try {
      const data = await getResident(id);
      setPerson(data);
      setLoading(false);
      setError(false);
    } catch (e) {
      setPerson(null);
      setLoading(false);
      setError(true);
    }
  };
  useEffect(() => {
    getPerson();
  }, []);

  if (loading) {
    return <Spinner />;
  }
  return (
    <>
      {error && <ErrorMessage />}
      {person && (
        <>
          <div className={cx('govuk-grid-row')}>
            <Select
              govGrid="one-half"
              name="category"
              options={category}
              label="Choose a form category"
              placeHolder="Choose one"
              onChange={(value) => setCategoryValue(value)}
            />
            {categoryValue && (
              <Select
                govGrid="one-half"
                name="formList"
                options={forms.filter(
                  (form) => form.category === categoryValue
                )}
                label="Choose a form"
                placeHolder="Choose one"
                onChange={(value) => setUrl(value)}
              />
            )}
          </div>
          {url && (
            <LinkButton
              label="Load form"
              route={
                ageContext === 'C'
                  ? `${url}${populateChildForm(
                      person.firstName,
                      person.lastName,
                      person.mosaicId,
                      user.name,
                      url
                    )}`
                  : url
              }
              internalQuery={`?id=${id}`}
            />
          )}
        </>
      )}
    </>
  );
};

AddForm.propTypes = {
  id: PropTypes.string.isRequired,
};

export default AddForm;
