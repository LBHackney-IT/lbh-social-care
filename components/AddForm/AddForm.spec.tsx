import { render, fireEvent, act } from '@testing-library/react';

import AddForm from './AddForm';
import { UserContext } from 'components/UserContext/UserContext';
import { mockedUser } from 'factories/users';
import { residentFactory } from 'factories/residents';

jest.mock('data/googleForms/adultForms', () => [
  {
    text: 'Foo - Adult',
    value: 'https://foo.com',
    category: 'General',
    id: 'foo_adult',
  },
  {
    text: 'Bar - Adult',
    value: 'https://foo.com',
    category: 'General',
  },
]);

jest.mock('data/googleForms/childForms', () => [
  {
    text: 'Foo - Child',
    value: 'https://foo.com',
    category: 'General',
  },
  {
    text: 'Bar - Child',
    value: 'https://foo.com',
    category: 'General',
  },
]);

describe('AddForm component', () => {
  it('should render adult forms', async () => {
    const props = {
      person: residentFactory.build({ contextFlag: 'A' }),
    };
    const { getByTestId, getByRole } = render(
      <UserContext.Provider
        value={{
          user: mockedUser,
        }}
      >
        <AddForm {...props} />
      </UserContext.Provider>
    );

    const autocompleteInput = getByTestId('formList') as HTMLInputElement;

    expect(autocompleteInput).toBeInTheDocument();
    await act(async () => {
      fireEvent.click(autocompleteInput);
    });
    expect(getByRole('option', { name: 'Case Note Recording' })).toBeDefined();

    await act(async () => {
      fireEvent.click(getByRole('option', { name: 'Bar - Adult' }));
    });
    expect(autocompleteInput.value).toBe('Bar - Adult');
  });

  it('should render children forms', async () => {
    const props = {
      person: residentFactory.build({ contextFlag: 'C' }),
    };
    const { getByRole, asFragment, getByTestId } = render(
      <UserContext.Provider
        value={{
          user: mockedUser,
        }}
      >
        <AddForm {...props} />
      </UserContext.Provider>
    );

    const autocompleteInput = getByTestId('formList') as HTMLInputElement;

    expect(autocompleteInput).toBeInTheDocument();
    await act(async () => {
      fireEvent.click(autocompleteInput);
    });
    expect(asFragment()).toMatchSnapshot();

    await act(async () => {
      fireEvent.click(getByRole('option', { name: 'Foo - Child' }));
    });
    expect(autocompleteInput.value).toBe('Foo - Child');
  });
});
