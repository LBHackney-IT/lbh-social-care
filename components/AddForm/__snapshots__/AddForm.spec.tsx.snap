// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddForm component should render children forms 1`] = `
<DocumentFragment>
  <div
    class="govuk-form-group"
  >
    <div
      aria-expanded="true"
      aria-haspopup="listbox"
      aria-labelledby="downshift-1-label"
      aria-owns="downshift-1-menu"
      class="govuk-form-group lbh-form-group"
      role="combobox"
    >
      <label
        class="govuk-label lbh-label--m"
        id="downshift-1-label"
      >
        Choose a form 
      </label>
      <div
        class="inputContainer"
      >
        <input
          aria-autocomplete="list"
          aria-controls="downshift-1-menu"
          aria-labelledby="downshift-1-label"
          autocomplete="off"
          class="govuk-input lbh-input govuk-input--width-20 "
          data-testid="formList"
          id="formList"
          placeholder="Select or type form name"
          value=""
        />
        <button
          aria-haspopup="true"
          aria-label="toggle menu"
          class="button"
          data-toggle="true"
          role="button"
          type="button"
        >
          <svg
            fill="none"
            height="10"
            viewBox="0 0 17 10"
            width="17"
          >
            <path
              d="M2 1.5L8.5 7.5L15 1.5"
              stroke="#0B0C0C"
              stroke-width="3"
            />
          </svg>
        </button>
      </div>
      <ul
        aria-labelledby="downshift-1-label"
        class="suggestionList"
        id="downshift-1-menu"
        role="listbox"
        style="border-bottom: 1px solid #aaa;"
      >
        <li
          aria-selected="false"
          class="suggestion"
          data-testid="formList_0"
          id="downshift-1-item-0"
          role="option"
          style="color: rgb(0, 0, 0); background-color: rgb(255, 255, 255);"
        >
          Bar - Child
        </li>
        <li
          aria-selected="false"
          class="suggestion"
          data-testid="formList_1"
          id="downshift-1-item-1"
          role="option"
          style="color: rgb(0, 0, 0); background-color: rgb(255, 255, 255);"
        >
          Foo - Child
        </li>
      </ul>
    </div>
  </div>
  <button
    class="govuk-button"
    data-module="govuk-button"
  >
    Load form
  </button>
</DocumentFragment>
`;
